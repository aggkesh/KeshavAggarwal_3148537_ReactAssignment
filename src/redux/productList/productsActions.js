import axios from 'axios'
import { 
    FETCH_PRODUCTS_REQUEST, 
    FETCH_PRODUCTS_SUCCESS, 
    FETCH_PRODUCTS_FAILURE,  
    LOW_TO_HIGH_SORT, 
    HIGH_TO_LOW_SORT, 
    HEADER_TOTAL_COUNT, 
    URL_PRODUCTS_FETCH,
    PAGE_QUERY,
    DEFAULT_LIMIT_QUERY,
    SORT_PRICE_QUERY,
    ORDER_QUERY,
    ASC_QUERY,
    DESC_QUERY,
    TITLE_LIKE_QUERY
} from "./productsTypes"

export const fetchProductsRequest = () => {
    return {
        type: FETCH_PRODUCTS_REQUEST
    }
}

export const fetchProductsSuccess = (products, productTotalCount, search, activePage, sort) => {
    return {
        type: FETCH_PRODUCTS_SUCCESS,
        productTotalCount: productTotalCount,
        search: search,
        activePage: activePage,
        sort: sort,
        payload: products
    }
}

export const fetchProductsFailure = error => {
    return {
        type: FETCH_PRODUCTS_FAILURE,
        payload: error
    }
}

export const fetchProducts = (search, activePage, sort, history) => {
    return (dispatch) => {
        createRequest(requestUrl(search, activePage, sort), search, activePage, sort, history, dispatch)
    }
}

const requestUrl = (search, activePage, sort) => {

    var url = baseURL(search) + PAGE_QUERY + activePage + '&' + DEFAULT_LIMIT_QUERY

    switch(sort) { 
        case LOW_TO_HIGH_SORT:
            url =  url + '&' + SORT_PRICE_QUERY + '&' + ORDER_QUERY + ASC_QUERY;
            break;
        case HIGH_TO_LOW_SORT:
            url =  url + '&' + SORT_PRICE_QUERY + '&' + ORDER_QUERY + DESC_QUERY;
            break;
        default:    
    }
    return url
}

const baseURL = (search) => {
    return search && search === '' ? URL_PRODUCTS_FETCH + '?' 
        : URL_PRODUCTS_FETCH + '?' + TITLE_LIKE_QUERY + search + '&';
}

const createRequest= (url, search, activePage, sort, history, dispatch) => {
    dispatch(fetchProductsRequest());    
    axios.get(url)
        .then(response => {
            const products = response.data
            dispatch(fetchProductsSuccess(products, 
                response.headers[HEADER_TOTAL_COUNT], search, activePage, sort))
        })
        .catch(error => {
            const errorMessage = error.message
            dispatch(fetchProductsFailure(errorMessage))
            history.push('/error/500')
        });
}