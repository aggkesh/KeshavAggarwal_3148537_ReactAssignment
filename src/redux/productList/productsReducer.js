import { 
    FETCH_PRODUCTS_REQUEST, 
    FETCH_PRODUCTS_SUCCESS, 
    FETCH_PRODUCTS_FAILURE,
    DEFAULT_SORT
} from "./productsTypes"

const initialState = {
    default: true,
    loading: false,
    search: '',
    activePage: 1,
    sort: DEFAULT_SORT,
    productTotalCount: 0,
    products: [],
    error: ''
}

const reducer = (state = initialState, action) => {
    switch (action.type) {
        case FETCH_PRODUCTS_REQUEST:
            return {
                ...state,
                default: false,
                loading: true
            }
        case FETCH_PRODUCTS_SUCCESS:
            return {
                ...state,
                loading: false,
                productTotalCount: action.productTotalCount,
                search: action.search,
                activePage: action.activePage,
                sort: action.sort,
                products: action.payload,
                error: ''
            }
        case FETCH_PRODUCTS_FAILURE:
            return {
                ...state,
                loading: false,
                products: [],
                error: action.payload
            }        
        default: return state    
    }
}

export default reducer