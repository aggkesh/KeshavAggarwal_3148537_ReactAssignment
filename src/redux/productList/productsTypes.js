export const FETCH_PRODUCTS_REQUEST = 'FETCH_PRODUCTS_REQUEST'
export const FETCH_PRODUCTS_SUCCESS = 'FETCH_PRODUCTS_SUCCESS'
export const FETCH_PRODUCTS_FAILURE = 'FETCH_PRODUCTS_FAILURE'
export const DEFAULT_SORT = 'Default'
export const HIGH_TO_LOW_SORT = 'High To Low'
export const LOW_TO_HIGH_SORT = 'Low To High'
export const URL_PRODUCTS_FETCH = 'http://localhost:3001/products'
export const PAGE_QUERY = '_page='
export const DEFAULT_LIMIT_QUERY = '_limit=8'
export const TITLE_LIKE_QUERY = 'title_like='
export const SORT_PRICE_QUERY = '_sort=price'
export const ORDER_QUERY = '_order='
export const ASC_QUERY = 'asc'
export const DESC_QUERY = 'desc'
export const HEADER_TOTAL_COUNT = 'x-total-count'