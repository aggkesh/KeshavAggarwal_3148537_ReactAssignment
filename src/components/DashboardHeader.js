import React from 'react';
import { Dropdown, Col } from 'react-bootstrap';
import { DEFAULT_SORT, LOW_TO_HIGH_SORT, HIGH_TO_LOW_SORT } from '../redux/productList/productsTypes';
import { useHistory  } from 'react-router-dom';

/**
 * Dashboard Header Ui 
 * 
 * @param {productData} product Data to show dashboard ui
 * @param {fetchProducts} function to fetch product from the store
 */
const DashboardHeader = ({ fetchProducts, productData }) => {
    const history = useHistory();
    const items = [DEFAULT_SORT, LOW_TO_HIGH_SORT, HIGH_TO_LOW_SORT];

    return (
        <React.Fragment>
                <Col>
                    <span className="display-4 font-weight-bold">Dashboard</span>
                </Col>
                <Col>
                    <Dropdown  className="float-right">
                        <Dropdown.Toggle variant="primary" id="dropdown-basic">
                            Sort
                        </Dropdown.Toggle>

                        <Dropdown.Menu>
                            {
                                items.map(item => <Dropdown.Item key={ item } active={ productData.sort === item } onClick={
                                        () => fetchProducts(productData.search, productData.activePage, item, history)}>{ item }
                                    </Dropdown.Item>)
                            }
                        </Dropdown.Menu>
                    </Dropdown>
                </Col>
        </React.Fragment>
    )
}

export default DashboardHeader