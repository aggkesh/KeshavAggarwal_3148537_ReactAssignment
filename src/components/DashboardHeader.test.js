import React from 'react';
import DashboardHeader from './DashboardHeader';

import Enzyme, { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { DEFAULT_SORT } from '../redux/productList/productsTypes';

Enzyme.configure({ adapter: new Adapter() })

describe('Dashboard Header Snapshot Test Cases', () => {
    it('renders correctly enzyme', () => {
        const productData = {
            activePage: 2,
            sort: DEFAULT_SORT,
            products: [
                {
                  id: 1,
                  title: "Google Pixel - Black",
                  img: "/img/product-1.png",
                  price: 19999,
                  info: "Lorem ipsum dolor amet offal butcher quinoa sustainable gastropub, echo park actually green juice sriracha paleo. Brooklyn sriracha semiotics, DIY coloring book mixtape craft beer sartorial hella blue bottle. Tote bag wolf authentic try-hard put a bird on it mumblecore. Unicorn lumbersexual master cleanse blog hella VHS, vaporware sartorial church-key cardigan single-origin coffee lo-fi organic asymmetrical. Taxidermy semiotics celiac stumptown scenester normcore, ethical helvetica photo booth gentrify."
                },
                {
                  id: 2,
                  title: "Samsung S7",
                  img: "/img/product-2.png",
                  price: 16,
                  info: "Lorem ipsum dolor amet offal butcher quinoa sustainable gastropub, echo park actually green juice sriracha paleo. Brooklyn sriracha semiotics, DIY coloring book mixtape craft beer sartorial hella blue bottle. Tote bag wolf authentic try-hard put a bird on it mumblecore. Unicorn lumbersexual master cleanse blog hella VHS, vaporware sartorial church-key cardigan single-origin coffee lo-fi organic asymmetrical. Taxidermy semiotics celiac stumptown scenester normcore, ethical helvetica photo booth gentrify."
                },
            ]
        }
        const wrapper = shallow(<DashboardHeader productData={ productData } />)
      
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});