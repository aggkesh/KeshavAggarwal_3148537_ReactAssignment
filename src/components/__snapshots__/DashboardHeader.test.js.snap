// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Header Snapshot Test Cases renders correctly enzyme 1`] = `
<Fragment>
  <Col>
    <span
      className="display-4 font-weight-bold"
    >
      Dashboard
    </span>
  </Col>
  <Col>
    <Dropdown
      className="float-right"
      navbar={false}
    >
      <DropdownToggle
        id="dropdown-basic"
        variant="primary"
      >
        Sort
      </DropdownToggle>
      <DropdownMenu
        alignRight={false}
        flip={true}
      >
        <DropdownItem
          active={true}
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "SafeAnchor",
              "render": [Function],
            }
          }
          disabled={false}
          key="Default"
          onClick={[Function]}
        >
          Default
        </DropdownItem>
        <DropdownItem
          active={false}
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "SafeAnchor",
              "render": [Function],
            }
          }
          disabled={false}
          key="Low To High"
          onClick={[Function]}
        >
          Low To High
        </DropdownItem>
        <DropdownItem
          active={false}
          as={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "SafeAnchor",
              "render": [Function],
            }
          }
          disabled={false}
          key="High To Low"
          onClick={[Function]}
        >
          High To Low
        </DropdownItem>
      </DropdownMenu>
    </Dropdown>
  </Col>
</Fragment>
`;
